### PR TITLE
[#9564] Consolidate PUT /students RESTFul API

### PR DESCRIPTION
**Pull request by Adoby7**
Part of #9564 

**Outline of Solution**
* Use request DTO 
* Rename the old endpoint to `PUT /students`
* Backend only returns a list of successful enrolled students. Logic of classifying new students, modified students,  unchanged students are delegated to frontend.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java [hunk 1]
```diff
@@ -1,31 +1,27 @@
...
+import teammates.ui.webapi.output.StudentsData;
+import teammates.ui.webapi.request.StudentsEnrollRequest;
 
 /**
- * Action: saving the list of enrolled students for a course of an instructor.
+ * Enroll a list of students.
+ * Update the students who are already existed.
+ * Returns all students who are successfully enrolled.
  */
 public class EnrollStudentsAction extends Action {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java]: Try to write a well-formatted Javadoc:
```
/**
* Enroll xxx
* 
* <p>Update xxxx
* <p>Create xxxx
* <p>Returns all xxxx
*/
```

File: src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java [hunk 2]
```diff
@@ -45,148 +41,56 @@ public void checkSpecificAccessControl() {
...
-            return new JsonResult(e.getMessage(), HttpStatus.SC_BAD_REQUEST);
-
-        } catch (EntityAlreadyExistsException e) {
-            log.severe("Entity already exists exception occurred when updating student: " + e.getMessage());
-
-            return new JsonResult("The enrollment failed, possibly because some students were re-enrolled before "
-                            + "the previous enrollment action was still being processed by TEAMMATES database "
-                            + "servers. Please try again after about 10 minutes. If the problem persists, "
-                            + "please contact TEAMMATES support", HttpStatus.SC_CONFLICT);
+        if (!logic.isCoursePresent(courseId)) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java]: You may not want to include this as `checkSpecificAccessControl` ensure the present of the course.

File: src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    public List<StudentEnrollRequest> getStudentEnrollRequests() {
+        return studentEnrollRequests;
+    }
+
+    @Override
+    public void validate() {
+        assertTrue(!studentEnrollRequests.isEmpty(), ERROR_MESSAGE_EMPTY_ENROLLMENT);
+        assertTrue(studentEnrollRequests.size() <= 100, ERROR_MESSAGE_TOO_MANY_ENROLLMENTS);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: magic number 100? I think there is a constant in `Const.java`

File: src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        Set<String> emails = new HashSet<>();
+        for (StudentEnrollRequest request : studentEnrollRequests) {
+            assertTrue(!emails.contains(request.getEmail()),
+                    String.format(ERROR_MESSAGE_DUPLICATE_EMAIL, request.getEmail()));
+            emails.add(request.getEmail());
+        }
+
+        Map<String, String> teamInSection = new HashMap<>();
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: There is a problem of putting the team validation logic here.

The teams that are already enrolled is not valided together with the students to enrol.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 5]
```diff
@@ -95,73 +94,178 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    }, (resp: ErrorMessageOutput) => {
+      this.statusMessage.pop(); // removes any existing error status message
+      this.statusMessageService.showErrorMessage(resp.error.message);
+    });
+    this.studentService.getStudentsFromCourse(this.courseid).subscribe((resp: Students) => {
+      this.existingStudents = resp.students;
+    });
+  }
+
+  private propagateEnrollResultPanelList(existingStudents: Student[], enrolledStudents: Student[],
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: populate is a better word :)

Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Javadoc?

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 6]
```diff
@@ -95,73 +94,178 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+                                         enrollRequests: StudentEnrollRequest[]): EnrollResultPanel[] {
+    const updateStatus: { [key: string]: number } = {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Let define this as an enum above

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 7]
```diff
@@ -95,73 +94,178 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
-          this.statusMessageService.showSuccessMessage('Enrollment successful. Summary given below.');
-          this.enrollResultPanelList = resp.enrollResultPanelList;
-        }, (resp: ErrorMessageOutput) => {
-          this.statusMessage.pop(); // removes any existing error status message
-          this.statusMessage.push({
-            message: resp.error.message,
-            color: 'danger',
-          });
+    const panelClass: { [key: number]: string } = {
+      0: 'bg-primary',
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Expose the enum to the HTML file and put the logic of choosing different CSS class in the HTML file.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 8]
```diff
@@ -95,73 +94,178 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    const studentLists: Student[][] = [];
+
+    for (const _ of Object.values(updateStatus)) {
+      studentLists.push([]);
+    }
+
+    // Identify students not in the enroll list.
+    for (const existingStudent of existingStudents) {
+      let isModified: boolean = false;
+      for (const enrolledStudent of enrolledStudents) {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: There is a function called find: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java (hunk 4 above)
Comment (Adoby7) [reply src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: Yes, and I put that part of validation in frontend.

Should I merge two parts of validation together and remove the validation here? As I think this validation here is for the request itself, I keep both currently

Comment (Adoby7) [reply src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: I see, for security reasons, I will put validation into backend

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.html [hunk 9]
```diff
@@ -261,7 +261,7 @@ <h1>Enrollment Results for {{ courseid }}</h1>
   </div>
   <div *ngFor="let enrollResultPanel of enrollResultPanelList">
     <div class="card enroll-results-panel" *ngIf="enrollResultPanel.studentList.length > 0">
-      <div class="card-header {{ enrollResultPanel.panelClass }}">
+      <div class="card-header {{ panelClass[enrollResultPanel.status] }}">
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.html]: Try ngClass https://angular.io/api/common/NgClass.
```suggestion
      <div class="card-header", [ngClass]="{'bg-primary': enrollResultPanel.status === EnrollStatus.NEW, 'bg-warning': enrollResultPanel.status === EnrollStatus.MODIFIED}">
```

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 10]
```diff
@@ -74,8 +74,10 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
   hotRegisterer: HotTableRegisterer = new HotTableRegisterer();
   newStudentsHOT: string = 'newStudentsHOT';
 
-  enrollData?: string;
+  // enum
+  panelClass: typeof PanelClass = PanelClass;
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: BTW, for enum it should be defined on the top of the class. The name should be `PanelClass` since it is more like a constant.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 11]
```diff
@@ -95,73 +98,139 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    // Parse the user input to be requests.
+    for (const row of spreadsheetData) {
+      const request: StudentEnrollRequest = {
+        section: row[hotInstanceColHeaders.indexOf(this.colHeaders[0])],
+        team: row[hotInstanceColHeaders.indexOf(this.colHeaders[1])],
+        name: row[hotInstanceColHeaders.indexOf(this.colHeaders[2])],
+        email: row[hotInstanceColHeaders.indexOf(this.colHeaders[3])],
+        comments: row[hotInstanceColHeaders.indexOf(this.colHeaders[4])],
+      };
+      if (request.comments === null) {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Will comments be `null`?

If there is a null-check for comments, how about for others?

They won't be null because of https://github.com/TEAMMATES/teammates/pull/9647/files#diff-cc2421f6f5575039136568f60626b85fR108?

File: src/web/services/student.service.ts [hunk 12]
```diff
@@ -15,12 +16,33 @@ export class StudentService {
...
    */
-  getStudentsFromCourse(courseId: string): Observable<Students> {
+  getStudentsFromCourse(courseId: string, teamName?: string): Observable<Students> {
+    if (teamName) {
+      const paramsMap: { [key: string]: string } = {
+        courseid: courseId,
+        teamname: teamName,
+      };
+      return this.httpRequestService.get('/students', paramsMap);
+    }  {
```
Comment (xpdavid) [inline src/web/services/student.service.ts]:
```suggestion
    } else {
```
?

File: src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * The request for enrolling a list of students.
+ */
+public class StudentsEnrollRequest extends BasicRequest {
+
+    private static final String ERROR_MESSAGE_DUPLICATE_EMAIL =
+            "Error, duplicated email addresses detected in the input: %s";
+    private static final String ERROR_MESSAGE_TOO_MANY_ENROLLMENTS = "Please enroll at most 100 students each time.";
+    private static final String ERROR_MESSAGE_EMPTY_ENROLLMENT = "Please input at least one student detail.";
```
[the remark this answers is not in the document]
Comment (Adoby7) [reply src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: However, the screenshot you show is marked as deprecated, so I don't think we should use it.

[the remark this answers is not in the document]
Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/request/StudentsEnrollRequest.java]: Nope. I mean you should make use the text. No the variable itself :P

File: src/web/services/student.service.ts (hunk 12 above)
Comment (Adoby7) [reply src/web/services/student.service.ts]: I tried to add "else" but then `npm lint` removes it 😅

File: src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+    public void testExecute_withNewStudent_shouldBeAddedToDatabase() {
+        String courseId = typicalBundle.students.get("student1InCourse1").getCourse();
+        StudentAttributes newStudent = getTypicalNewStudent(courseId);
+        StudentsEnrollRequest req = prepareRequest(Arrays.asList(newStudent));
+
+        loginAsInstructor(typicalBundle.instructors.get("instructor1OfCourse1").getGoogleId());
+        List<StudentData> enrolledStudents = executeActionAndReturnResults(courseId, req);
+
+        assertEquals(1, enrolledStudents.size());
+        verifyStudentInDatabase(newStudent, enrolledStudents.get(0).getCourseId(), enrolledStudents.get(0).getEmail());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java]: You should verify at least have one test case verifying `enrolledStudents` has correct content?

File: src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+        loginAsInstructor(typicalBundle.instructors.get("instructor1OfCourse1").getGoogleId());
+        List<StudentData> enrolledStudents = executeActionAndReturnResults(courseId, req);
+
+        assertEquals(1, enrolledStudents.size());
+        verifyStudentInDatabase(studentToUpdate, enrolledStudents.get(0).getCourseId(), enrolledStudents.get(0).getEmail());
+
+    }
+
+    @Test
+    public void testExecute_withEmailFieldChanged_shouldCreateNewStudent() {
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java]: Is this the same as `testExecute_withNewStudent_shouldBeAddedToDatabase`?

File: src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        invalidNewStudent.email = "invalidEmail";
+        StudentAttributes validExistingStudent = typicalBundle.students.get("student1InCourse1");
+        validExistingStudent.name = "new name";
+        StudentAttributes invalidExistingStudent = typicalBundle.students.get("student2InCourse1");
+        invalidExistingStudent.team = "invalid | team % name";
+        StudentsEnrollRequest req = prepareRequest(
+                Arrays.asList(validNewStudent, invalidNewStudent, validExistingStudent, invalidExistingStudent));
+
+        loginAsInstructor(typicalBundle.instructors.get("instructor1OfCourse1").getGoogleId());
+        List<StudentData> enrolledStudents = executeActionAndReturnResults(courseId, req);
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java]: Same here. The `enrolledStudents` should be verified.

File: src/web/app/pages-instructor/instructor-course-enroll-page/enroll-status.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+  /**
+   * The student exists in the course, and some fields are modified.
+   */
+  MODIFIED = 1,
+
+  /**
+   * The student exists in the course, and nothing is changed during the enrollment.
+   */
+  MODIFIEDUNCHANGED = 2,
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-enroll-page/enroll-status.ts]:
```suggestion
  MODIFIED_UNCHANGED = 2,
```

**Comment by xpdavid** [comment]
On a related note, this will break the current LNP test. I have a quick look at it and it seems that there is only little modification needed. @jacoblipech, can you also help @Adoby7 do this?

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 18]
```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
-
   /**
    * Submits enroll data
    */
   submitEnrollData(): void {
     const newStudentsHOTInstance: Handsontable =
         this.hotRegisterer.getInstance(this.newStudentsHOT);
-    const spreadsheetData: string[][] = newStudentsHOTInstance.getData();
+    const spreadsheetData: string[][] = newStudentsHOTInstance.getData()
+        .filter((row: string[]) => (!row.every((cell: string) => cell === null || cell === '')));
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: How about this so we don't need the for loop in line 116?
```javascript suggestion
        newStudentsHOTInstance.getData()
        .filter((row: string[]) => (!row.every((cell: string) => cell === null || cell === '')))
        .forEach((row: string[]) => (studentsEnrollRequest.studentEnrollRequests.push({
          section: row[hotInstanceColHeaders.indexOf(this.colHeaders[0])] === null ?
            '' : row[hotInstanceColHeaders.indexOf(this.colHeaders[0])],
          team: row[hotInstanceColHeaders.indexOf(this.colHeaders[1])] === null ?
            '' : row[hotInstanceColHeaders.indexOf(this.colHeaders[1])],
          name: row[hotInstanceColHeaders.indexOf(this.colHeaders[2])] === null ?
            '' : row[hotInstanceColHeaders.indexOf(this.colHeaders[2])],
          email: row[hotInstanceColHeaders.indexOf(this.colHeaders[3])] === null ?
            '' : row[hotInstanceColHeaders.indexOf(this.colHeaders[3])],
          comments: row[hotInstanceColHeaders.indexOf(this.colHeaders[4])] === null ?
            '' : row[hotInstanceColHeaders.indexOf(this.colHeaders[4])],
        })));
```

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 19]
```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    });
+  }
+
+  private populateEnrollResultPanelList(existingStudents: Student[], enrolledStudents: Student[],
+                                         enrollRequests: StudentEnrollRequest[]): EnrollResultPanel[] {
+
+    const panels: EnrollResultPanel[] = [];
+    const studentLists: Student[][] = [];
+
+    for (const _ of Object.values(EnrollStatus).filter((value: EnrollStatus) => typeof value === 'string')) {
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: What's the purpose of this loop if we are always pushing 5 empty arrays (corresponds to the number of enroll types in `enroll-status.ts`?

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 20]
```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    for (const existingStudent of existingStudents) {
+      const enrolledStudent: Student | undefined = enrolledStudents.find((student: Student) => {
+        return student.email === existingStudent.email;
+      });
+      if (enrolledStudent === undefined) {
+        studentLists[EnrollStatus.UNMODIFIED].push(existingStudent);
+      }
+    }
+
+    // Identify new students, modified students, and modified without any change students.
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: > and modified without any change students.

and students that are modified without any changes.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 21]
```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+      if (unchangedStudent !== undefined) {
+        studentLists[EnrollStatus.MODIFIEDUNCHANGED].push(enrolledStudent);
+      } else if (unchangedStudent === undefined && modifiedStudent !== undefined) {
+        studentLists[EnrollStatus.MODIFIED].push(enrolledStudent);
+      } else if (unchangedStudent === undefined && modifiedStudent === undefined) {
+        studentLists[EnrollStatus.NEW].push(enrolledStudent);
+      }
+    }
+
+    // Identify unsuccessfully enrolled students.
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: <s>Identify unsuccessfully enrolled students.</s> 

Identify students that failed to enroll.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 22]
```diff
@@ -95,73 +97,143 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+        studentLists[EnrollStatus.ERROR].push({
+          email: request.email,
+          courseId: this.courseid,
+          name: request.name,
+          sectionName: request.section,
+          teamName: request.team,
+          comments: request.comments,
+          joinState: JoinState.NOT_JOINED,
+          lastName: '',
+          requestId: '',
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Where are these used?

```
          joinState: JoinState.NOT_JOINED,
          lastName: '',
          requestId: '',
```

**Comment by tanhengyeow** [comment]
On a related note, are you going to fix #9663 in this PR?

**Comment by Adoby7** [comment]
@tanhengyeow I am not going to do #9663 here. I think it is better to do it when consolidating another API call (`get enrolledStudents`, which should be deprecated), as the response DTO should be changed

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts (hunk 22 above)
Comment (Adoby7) [reply src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Actually they are not used, but if I do not initialize these fields the compiler will throw error.

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts (hunk 19 above)
Comment (Adoby7) [reply src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Yes, the purpose here is just to push 5 empty arrays. Otherwise when I access something like `studentLists[3]` (which is `studentLists[EnrollStatus.ERROR]` in the code), it will throw error

File: src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java (hunk 15 above)
Comment (Adoby7) [reply src/test/java/teammates/test/cases/webapi/EnrollStudentsActionTest.java]: Here we also check that the original student, although with same `name` `team` `section` and `comment`, is not modified

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts (hunk 22 above)
Comment (Adoby7) [reply src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: On the separate note, I find `requestId` is not necessary so I have removed it

**Comment by Adoby7** [comment]
Hi, @tanhengyeow, could you look at it again? I think the error message is more meaningful now. The specification is that `Section` and `Comment` fields are optional while the other three are not

File: src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts [hunk 23]
```diff
@@ -95,73 +97,139 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
...
+    for (const status of Object.values(EnrollStatus).filter((value: EnrollStatus) => typeof value === 'string')) {
+      panels.push({
+        status: EnrollStatus[status as keyof typeof EnrollStatus],
+        messageForEnrollmentStatus: statusMessage[EnrollStatus[status as keyof typeof EnrollStatus]],
+        studentList: studentLists[EnrollStatus[status as keyof typeof EnrollStatus]],
+      });
+    }
+
+    if (studentLists[EnrollStatus.ERROR].length > 0) {
+      this.statusMessageService.showErrorMessage('Some students are failed to be enrolled, see summary below');
```
Comment (tanhengyeow) [inline src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts]: Some students <s>are</s> failed to be enrolled

**Comment by Adoby7** [comment]
@tanhengyeow Yes, the error message is a problem. As the API is consolidated to be RESTFul, the workflow has changed. Now if the invalid student does not affect other students (e.g. just an invalid email), backend will simply ignore invalid student and continue other enrollments. In addition, the API response contains students who are enrolled successfully, and I could not find a way to return both correct response and error response simultaneously. 

Therefore the way to solve it is to move such validation into frontend, and show the correct error message. However, I find this could contain large amount of migration, and would be better to work in a separate PR. What is your opinion about this?

File: src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+            studentsToEnroll.add(StudentAttributes.builder(courseId, studentEnrollRequest.getEmail())
+                    .withName(studentEnrollRequest.getName())
+                    .withSectionName(studentEnrollRequest.getSection())
+                    .withTeamName(studentEnrollRequest.getTeam())
+                    .withComment(studentEnrollRequest.getComments())
+                    .build());
+        });
+
+        List<StudentAttributes> existingStudents = logic.getStudentsForCourse(courseId);
+        validateTeamName(existingStudents, studentsToEnroll);
```
Comment (junming403) [inline src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java]: So In here if one of the students to enroll is not valid, the entire request will fail?

Comment (Adoby7) [reply src/main/java/teammates/ui/webapi/action/EnrollStudentsAction.java]: Yes, because this error makes this student conflict with other students. If the invalid student does not affect others (e.g. the email does not match regex), it will not stop the entire request

**Comment by Adoby7** [comment]
@tanhengyeow I have added some general error message in the prompted window. Probably this would help. If we want the detailed error message, probably it needs a separate PR to do

**Comment by tanhengyeow** [comment]
@Adoby7 Reminder to fix the lint error :)

File: src/main/java/teammates/common/util/Const.java [hunk 25]
```diff
@@ -56,7 +56,7 @@
 
     public static final String ACTION_RESULT_FAILURE = "Servlet Action Failure";
 
-    public static final int SIZE_LIMIT_PER_ENROLLMENT = 150;
+    public static final int SIZE_LIMIT_PER_ENROLLMENT = 100;
```
Comment (damithc) [inline src/main/java/teammates/common/util/Const.java]: Why is this changed to 100? I think the 150 was intentional, to allow courses with just over 100 students to be enrolled in one shot.

Comment (Adoby7) [reply src/main/java/teammates/common/util/Const.java]: image
The enroll pages says maximum 100 students in one enrollment. Do we change it back?

Comment (damithc) [reply src/main/java/teammates/common/util/Const.java]: It was a hidden feature. The advertised limit remains 100. Change it back to 150 and put a comment to explain why it is 150?

Comment (Adoby7) [reply src/main/java/teammates/common/util/Const.java]: So, do we show the users maximum enroll size is 100, but if they try to enroll students slightly more than 100, we still allow it?

Comment (wkurniawan07) [reply src/main/java/teammates/common/util/Const.java]: @damithc the bigger question is, why does it become a "hidden" feature?

Comment (damithc) [reply src/main/java/teammates/common/util/Const.java]: 100 is a good round number and easy for users too e.g., when dividing a spreadsheet into batches.
However, it turns out that it is fairly common for instructors to have just over 100 students and also fairly common for them to not see the limit and try to enroll all at once. Allowing a buffer of 50 more allow such requests to go through -> reduces annoyance to users and saves us some requests too. This is a bit similar to having a hidden buffer for submission deadlines. 
Having said that, I just realized this hidden feature is no longer usable because the spreadsheet limits the number of rows to 100. Which means we can go back to 100. Just make sure there is no off-by-one error. i.e., we support 100, not 99, students at one go (I remember having a bug like that once)

Comment (Adoby7) [reply src/main/java/teammates/common/util/Const.java]: OK

**Comment by Adoby7** [comment]
Hi @wkurniawan07, could you look at this PR again, as there were some conflicts after your approval. Thank you

**Comment by amrut-prabhu** [comment]
@Adoby7 If you're referring to the [conflicts in commit 7318cd9](https://github.com/TEAMMATES/teammates/pull/9647/commits/7318cd9f36a3e5ddac36c99c7eca0f5586cca91b), the changes look fine to me
